### DISCUSSION TITLE
Role of the .usr file in NekRS

### DISCUSSION THREAD

**Discussion by JunaidAhmed2520** [Nek5000/nekRS#572] [Q&A]
Hi, I am trying to transition from Nek5000 and the usage of the .usr file is not clear to me having gone through the examples of conj_ht, channel and mv_cyl. The following are some of the questions that I would like to ask:

0. Should the .usr file always be defined?
1. Should the subroutines in the .usr files be called separately or is it done automatically?
2. What is the role of the `useric` subroutine while setting the initial conditions?
3. In the case of `mv_cyl`, I had observed that both the .usr and .oudf files contain the `uservp` function which sets the same variable properties but according to their own rules. Should both the .usr and .oudf files always  define the variable properties in the same way? What happens if they differ from each other?
4. Lastly, what are the steps involved in calculating torque and vorticity data using NekRS?
 
I would appreciate any help in this matter. Thanks in advance.

**Comment by yslan** [comment]
`.usr` file is the main user file in Nek5000. NekRS uses it as an interface to Nek5000 so we can still read Nek5000's case and reuse the setup with minimal modification.

0. No. You can run the case without an usr file. See this [example](https://github.com/Nek5000/nekRS/tree/master/examples/tcf)
1. At certain [location](https://github.com/Nek5000/nekRS/blob/master/src/nekInterface/nekInterface.f) in source code, those user subroutines (listed below) will be called.
   ```
      subroutine userchk   ! not called by default in repo/next
      subroutine useric (ix,iy,iz,ieg)
      subroutine usrdat3
      subroutine usrdat2
      subroutine usrdat0
      subroutine usrdat
      subroutine userservert
   ```
2. useric sets initial condition, but you can set it inside `.udf` as well. See tcf example.
3. `uservp` in `.usr` won't be called at all. The one in `.udf` is the main one. 
   Same as `userf` and `userq`.    I guess sometimes we keep them in usr as a dummy in case we would like to run the same case in Nek5000.
   The `userbc` is split into device functions in `.oudf`

4. For voracity, see this v23 example
   https://github.com/Nek5000/nekRS/blob/4f87e0e2ec0492e2a3edf27791252d6886814d00/examples/tgv/tgv.udf#L28-L34
    For torque, v23 has drag computation in this example 
   https://github.com/Nek5000/nekRS/blob/4f87e0e2ec0492e2a3edf27791252d6886814d00/examples/ktauChannel/ci.inc#L65

**Reply by JunaidAhmed2520** [reply]
Hi, thanks for the reply. 
I realise that there are various other variables that the `nrs` structure contains apart from the ones mentioned in the [read the docs](https://nekrsdoc.readthedocs.io/en/latest/index.html) website. As a user who is only aware of the information present on this website (it doesn't contain enough information about the `platform` structure), from where can I get to know more about the various available functions that nekRS provides. Could you please suggest some resources that are available so that I can get a **complete idea** about them. (including the postProcessing namespace as well)
